### PR TITLE
Introduce SolverRelating type relation to the new solver

### PR DESCRIPTION
**Pull request by compiler-errors**
Redux of #128744.

Splits out relate for the new solver so that implementors don't need to implement it themselves.

r? lcnr

### CODE REVIEW

**Comment by compiler-errors** [comment]
> but until we get rid of LUB/GLB this won't be fully possible

We could also, perhaps, get rid of `CombineFields` and just call `infcx.at.eq` when needing to equate in LUB. I could do that as a follow-up, especially because it has some overhead with perf since we'd need to clone obligationcauses.

File: compiler/rustc_type_ir/src/relate/solver_relating.rs [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+            let tcx = self.cx();
+            let opt_variances = tcx.variances_of(item_def_id);
+            relate_args_with_variances(self, item_def_id, opt_variances, a_arg, b_arg, false)
+        }
+    }
+
+    fn relate_with_variance<T: Relate<I>>(
+        &mut self,
+        variance: ty::Variance,
+        _info: VarianceDiagInfo<I>,
```
Comment (lcnr) [inline compiler/rustc_type_ir/src/relate/solver_relating.rs:125]: which type relations still care about `VarianceDiagInfo` 🤔 can look into it myself later

Comment (compiler-errors) [reply compiler/rustc_type_ir/src/relate/solver_relating.rs:125]: nll relating i think

**Comment by compiler-errors** [comment]
Now that the other PR landed, I believe this is ready

File: compiler/rustc_infer/src/infer/relate/mod.rs [hunk 2]
```diff
@@ -3,12 +3,10 @@
 //! As well as the implementation of `Relate` for interned things (`Ty`/`Const`/etc).
 
 pub use rustc_middle::ty::relate::RelateResult;
-pub use rustc_next_trait_solver::relate::*;
-
-pub use self::combine::PredicateEmittingRelation;
+pub use rustc_type_ir::relate::combine::PredicateEmittingRelation;
+pub use rustc_type_ir::relate::*;
 
 #[allow(hidden_glob_reexports)]
```
Comment (lcnr) [inline compiler/rustc_infer/src/infer/relate/mod.rs:9]: ßßtfzf

File: compiler/rustc_infer/src/infer/relate/type_relating.rs [hunk 3]
```diff
@@ -56,15 +56,14 @@ impl<'infcx, 'tcx> TypeRelating<'infcx, 'tcx> {
         trace: TypeTrace<'tcx>,
         param_env: ty::ParamEnv<'tcx>,
         define_opaque_types: DefineOpaqueTypes,
-        structurally_relate_aliases: StructurallyRelateAliases,
         ambient_variance: ty::Variance,
     ) -> TypeRelating<'infcx, 'tcx> {
+        assert!(!infcx.next_trait_solver);
```
Comment (lcnr) [inline compiler/rustc_infer/src/infer/relate/type_relating.rs:61]:
```suggestion
        assert!(!infcx.next_trait_solver());
```

File: compiler/rustc_type_ir/src/relate/combine.rs [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+    fn register_predicates(
+        &mut self,
+        obligations: impl IntoIterator<Item: Upcast<I, I::Predicate>>,
+    );
+
+    /// Register `AliasRelate` obligation(s) that both types must be related to each other.
+    fn register_alias_relate_predicate(&mut self, a: I::Ty, b: I::Ty);
+}
+
+pub trait InferCtxtCombineExt<I: Interner>: InferCtxtLike {
```
Comment (lcnr) [inline compiler/rustc_type_ir/src/relate/combine.rs:42]: style nit, does not impact r=me, free functions instead?

File: compiler/rustc_type_ir/src/relate/solver_relating.rs [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+        TypeError<Self::Interner>,
+    > {
+        let mut relate =
+            SolverRelating::new(self, StructurallyRelateAliases::Yes, ty::Invariant, param_env);
+        relate.relate(lhs, rhs)?;
+        Ok(relate.goals)
+    }
+}
+
+#[allow(unused)]
```
Comment (lcnr) [inline compiler/rustc_type_ir/src/relate/solver_relating.rs:65]: ?

File: compiler/rustc_type_ir/src/relate/solver_relating.rs [hunk 6]
```diff
@@ -0,0 +1,401 @@
...
+            ty::Covariant => {
+                self.infcx.sub_regions(b, a);
+            }
+            // Suptype(&'a u8, &'b u8) => Outlives('b: 'a) => SubRegion('a, 'b)
+            ty::Contravariant => {
+                self.infcx.sub_regions(a, b);
+            }
+            ty::Invariant => {
+                self.infcx.equate_regions(a, b);
+            }
```
Comment (lcnr) [inline compiler/rustc_type_ir/src/relate/solver_relating.rs:254]: `ty::Covariant => self.infcx.sub_regions(b, a),`

**Comment by lcnr** [comment]
r=me after review

**Pull request merged**